csrc/scheduler/matmul_utils.cpp

```diff
@@ -294,6 +315,98 @@ std::string isMatmulFusionDefinitionSupported(
   return "";
 }
 
+// Assume that tens has a contiguous dimension, and that we will load rows of
+// that dimension, without merging with another dimension first. Then determine
+// the maximum vectorization that can be used.
+//
+// If the argument has no contiguous dimensions, then a vectorization width of 1
+// is returned.
+//
+// These rows can start at any multiple of the non-contiguous strides, so we
+// seek the largest power of 2 that divides all those other dimensions (capped
+// to 16) as well as the data pointer.
+int64_t maxRowVectorization(const at::Tensor& tens) {
+  const int64_t data_ptr_int =
+      static_cast<int64_t>(reinterpret_cast<std::uintptr_t>(tens.data_ptr()));
+  int64_t vec_size = scheduler_utils::maxVectorizationWidth(data_ptr_int);
+  vec_size = std::min(vec_size, (int64_t)(16l));
+  vec_size /= (int64_t)tens.element_size();
+  vec_size = std::max(vec_size, 1l);
+  if (vec_size == 1l) {
+    return vec_size;
+  }
+
+  std::vector<int64_t> strides = tens.strides().vec();
+  std::sort(strides.begin(), strides.end());
+  NVF_ERROR(!strides.empty());
+  if (strides.front() != 1l) {
+    // Tensor is discontiguous
+    return 1l;
+  }
+  strides.erase(strides.begin());
+  for (auto stride_i : strides) {
+    if (vec_size == 1l) {
+      break;
+    }
+    vec_size =
+        std::min(vec_size, scheduler_utils::maxVectorizationWidth(stride_i));
+  }
+
+  return vec_size;
+}
+
+MatmulParams::SupportedVectorization getSupportedVectorization(
+    const mma_utils::RolesMap& roles_map,
+    SchedulerRuntimeInfo& runtime_info) {
+  auto getMinVectorization = [&roles_map,
+                              &runtime_info](MatmulRole role) -> int64_t {
+    int64_t vec_size = 16; // max vectorization size
+    const auto it = roles_map.find(role);
+    if (it == roles_map.end()) {
+      return 16;
+    }
+    for (TensorView* tv : it->second) {
+      int64_t v = maxRowVectorization(
+          runtime_info.expressionEvaluator().evaluate(tv).as<at::Tensor>());
+      if (v < vec_size) {
+        vec_size = v;
+      }
+      if (v == 1) {
+        // No need to continue analyzing if we know we cannot vectorize
+        break;
+      }
+    }
+    return vec_size;
+  };
+  MatmulParams::SupportedVectorization supported_vec_size;
+  supported_vec_size.a = getMinVectorization(MatmulRole::INPUT_A);
+  supported_vec_size.b = getMinVectorization(MatmulRole::INPUT_B);
+  // If there is no C tensor, then we will use full vectorization of the
+  // outputs. So find the max vectorization of the output tensors, then
+  // possibly reduce it if it is not supported by a input C tensor.
+  supported_vec_size.epilogue = getMinVectorization(MatmulRole::INPUT_C);
```

Comment:
This is a great point. We should support as general an epilogue as possible and that includes transposes. I had implicitly assumed we would schedule each aux input like the epilogue since that's what we currently do:
https://github.com/NVIDIA/Fuser/blob/eb75ab282dfe285cb8db7bcd5f8470f07ec889bb/csrc/scheduler/matmul.cpp#L647-L650
Perhaps we could use an issue to track at least that case of a transposed 2D bias. I could imagine eventually supporting normalizations and reductions in the epilogue as well, for example a loss function in an MLP.

Regarding the options you presented:

1. I believe so far the epilogue is ignored entirely in `canScheduleCompileTime`: https://github.com/NVIDIA/Fuser/blob/eb75ab282dfe285cb8db7bcd5f8470f07ec889bb/csrc/scheduler/matmul_utils.cpp#L313
All that is checked is that there is a mul-sum pattern or `MmaOp`. That function guards which fusions we will optionally accept as nvfuser matmuls once we have a new ATen matmul scheduler (soon). Since we know there are cases that are likely to fail (transposes, reductions, normalizations) we should at least filter those out. I think it makes sense to lock this down to only pointwise ops with inputs that have the same stride order as the output. The stride order requirement could be lifted in a separate PR.
2. We should definitely have a complete vectorization analysis accepting a broader collection of epilogues. I will have a look at #807 again now that I'm more familiar with this problem.

I see now that epilogue vectorization is not so simple. As you say, it does make sense to move it to a separate PR. I will strip the current PR down to just inferring A and B vectorization, leaving epilogue vectorization set to the maximum supported by the dtype. I'll file an issue to track epilogue analysis.

Reply:
Opened issues #2167 and #2169